presto-hive/src/main/java/com/facebook/presto/hive/parquet/ParquetHiveRecordCursor.java

```diff
@@ -771,78 +780,183 @@ public void end()
                 return new ParquetPrimitiveConverter(prestoType, fieldIndex);
             }
         }
 
-        return createGroupConverter(prestoType, columnName, parquetType, fieldIndex);
+        return createGroupConverter(prestoType, columnName, parquetType, fieldIndex, useNames, typeManager);
     }
 
-    private static GroupedConverter createGroupConverter(Type prestoType, String columnName, parquet.schema.Type parquetType, int fieldIndex)
+    private static GroupedConverter createGroupConverter(Type prestoType, String columnName, parquet.schema.Type parquetType, int fieldIndex, boolean useNames, TypeManager typeManager)
     {
         GroupType groupType = parquetType.asGroupType();
         switch (prestoType.getTypeSignature().getBase()) {
             case ARRAY:
-                return new ParquetListConverter(prestoType, columnName, groupType, fieldIndex);
+                return new ParquetListConverter(prestoType, columnName, groupType, fieldIndex, useNames, typeManager);
             case MAP:
-                return new ParquetMapConverter(prestoType, columnName, groupType, fieldIndex);
+                return new ParquetMapConverter(prestoType, columnName, groupType, fieldIndex, useNames, typeManager);
             case ROW:
-                return new ParquetStructConverter(prestoType, columnName, groupType, fieldIndex);
+                return new ParquetStructConverter(prestoType, columnName, groupType, fieldIndex, useNames, typeManager);
             default:
                 throw new IllegalArgumentException("Column " + columnName + " type " + parquetType.getOriginalType() + " not supported");
         }
     }
 
+    private static class Pair
+    {
+        final Type type;
+        final int index;
+
+        private Pair(int index, Type type)
+        {
+            this.index = index;
+            this.type = type;
+        }
+
+        static Pair of(int index, Type type)
+        {
+            return new Pair(index, type);
+        }
+    }
+
     private static class ParquetStructConverter
             extends GroupedConverter
     {
         private static final int NULL_BUILDER_POSITIONS_THRESHOLD = 100;
         private static final int NULL_BUILDER_SIZE_IN_BYTES_THRESHOLD = 32768;
 
         private final Type rowType;
-        private final int fieldIndex;
+        private int fieldIndex;
 
-        private final List<BlockConverter> converters;
+        private final List<BlockFieldConverter> converters;
         private BlockBuilder builder;
         private BlockBuilder nullBuilder; // used internally when builder is set to null
         private BlockBuilder currentEntryBuilder;
 
-        public ParquetStructConverter(Type prestoType, String columnName, GroupType entryType, int fieldIndex)
+        private int[] converterIndexes;
+        private List<Type> outOfOrderFieldTypes;
+        private InterleavedBlockBuilder outOfOrderBlockBuilder; // a tmp builder for fields out of order
```

Comment:
I don't see this `outOfOrderBlockBuilder` closed anywhere, not sure whether this is a problem though.